todo list, add, delete, edit, move features

Sena-dogan opened the pull request.

Sena-dogan committed:
Create a ListView

Add a ListView in this app's initial screen and also created ListRowView for ListView item

Sena-dogan committed:
Create AddView

Created an AddView for Add Button's Navigation destination

Sena-dogan committed:
data model for todo items

created a custom data model for todo items

Sena-dogan committed:
Add .gitignore file

Sena-dogan committed:
delete and move item functions

created basic delete and move item functions

Sena-dogan committed:
Created ListViewModel

move functions and items to the ListViewModel

Sena-dogan committed:
fix crash :))

Sena-dogan committed:
implemented functions to view

Sena-dogan committed:
Made suitable for mvvm structure

moved add view functions to the list view model

Sena-dogan committed:
fix crash again :)))

beyzaince commented on TodoBee/ViewModels/AddViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+class AddViewModel: ObservableObject {
+    
+    func deleteItem(indexSet: IndexSet) {
+        items.remove(atOffsets: indexSet)
```
where is the item definition? Are we sure about this code? When we run these project is it possible run ehe

Sena-dogan replied:
dosya silinmemişş sildimm :)))))

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,65 @@
+import Foundation
+import SwiftUI
+
+class ListViewModel: ObservableObject {
```
final

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,65 @@
+class ListViewModel: ObservableObject {
+    
+    @Published var items: [ItemModel] = []
+    @Published var alertTitle: String = ""
```
why did we define alert title with Published

Sena-dogan replied:
showAlert zaten published o yüzden burda published e gerek yok sanırımmm 👍 👍

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,65 @@
+    
+    func textIsAppropriate(text: String) -> Bool {
+        if text.count <= 0 {
+            alertTitle = "Your item can not be empty 😱😱😱"
```
constant

Sena-dogan replied:
Created a file for all constants 💯 💯

beyzaince commented on TodoBee/Views/AddView.swift:
```diff
@@ -0,0 +1,54 @@
+            .padding(14)
+        }
+        .navigationTitle("Add an Item ✍︎")
+        .alert(isPresented: $listViewModel.showAlert) {
```
should we toggle show alert function when we close these alert?

Sena-dogan replied:
created dismiss button

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    func saveButtonPressed(text: String, presentationMode: Binding<PresentationMode>) {
+        if textIsAppropriate(text: text) {
```
it would be better to use guard to control these logic instead of if

beyzaince commented on TodoBee/Views/ListRowView.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    var body: some View {
+        HStack {
+            Image(systemName: item.isCompleted ? "checkmark.circle" : "circle")
```
nice to have view model to move these logic

Sena-dogan replied:
created ListRowViewModel 👍 👍

beyzaince commented on TodoBee/Views/ListView.swift:
```diff
@@ -0,0 +1,46 @@
+            .onMove(perform: listViewModel.moveItem)
+        }
+        .listStyle(PlainListStyle())
+        .navigationTitle("Todo List ✔︎")
```
constant

beyzaince commented on TodoBee/Views/ListView.swift:
```diff
@@ -0,0 +1,46 @@
+                        }
+                    }
+            }
+            .onDelete(perform: listViewModel.deleteItem)
```
why did not we use item index i ?

Sena-dogan replied:
I used it noww :)) 👍

Sena-dogan committed:
fixes for reviews

Sena-dogan committed:
fix ListRowView call in the ListView

Sena-dogan committed:
save todo items with userdefaults

beyzaince commented on TodoBee/ViewModels/ListRowViewModel.swift:
```diff
@@ -0,0 +1,25 @@
+import Foundation
+import SwiftUI
+
+class ListRowViewModel: ObservableObject {
```
final

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+import Foundation
+import SwiftUI
+
+final class ListViewModel: ObservableObject {
```
😍 😍 😍 😍 finaaaaal 😍 😍 😍 😍 😍

Sena-dogan replied:
😎😎😋

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+    }
+    var alertTitle: String = ""
+    @Published var showAlert: Bool = false
+    let itemsKey: String = "items_list "
```
"items_list " ehe

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+    
+    func getItems() {
+        guard
+            let data = UserDefaults.standard.data(forKey: itemsKey),
```
it would be better to define `UserDefaultsHelper` class in here for user defaults logics

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+    
+    func addItem(title: String) {
+        let newItem = ItemModel(title: title, isCompleted: false)
+        items.append(newItem)
```
items.append(ItemModel(title: title, isCompleted: false))

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+    }
+    
+    func updateItem(at index: Int) {
+        items[index] = items[index].updateModel()
```
items[index].isCompleted.toggle()

beyzaince commented on TodoBee/ViewModels/ListViewModel.swift:
```diff
@@ -0,0 +1,74 @@
+    
+    func saveItems() {
+        if let encodeData = try? JSONEncoder().encode(items) {
+            UserDefaults.standard.set(encodeData, forKey: itemsKey)
```
.

Sena-dogan committed:
create user defaults helper

Sena-dogan committed:
Refactor addItem method

Moved item append logic inside addItem method for clarity

Sena-dogan committed:
Fix updateItem method by creating a mutable copy of the item

- Made isCompleted property mutable in ItemModel
- Updated updateItem method to toggle isCompleted on a mutable copy
